ambry-store/src/main/java/com/github/ambry/store/StorageManager.java

```diff
@@ -817,4 +826,165 @@ private void resumeDecommission(String partitionName) throws Exception {
+      // remove the replica from property store right away.
+      // StorageManager relies on state transition messages to add and remove replicasOnFailedDisks, if there is no state transition
+      // messages, then when we reset the partitions, replica list in the StorageManager would be obsolete. Fortunately
+      // all the replicasOnFailedDisks in the failed disks are already stopped, we just have to remove the disk from disk maps.
+      failedDisk.addAll(newFailedDisks);
+      long healthyDiskCapacity = diskToDiskManager.keySet()
+          .stream()
+          .filter(((Predicate<DiskId>) failedDisk::contains).negate())
+          .mapToLong(DiskId::getRawCapacityInBytes)
+          .sum();
+      List<ReplicaId> replicasOnFailedDisks = partitionNameToReplicaId.values()
+          .stream()
+          .filter(replica -> newFailedDisks.contains(replica.getDiskId()))
+          .collect(Collectors.toList());
+      logger.info("Replicas on the failed disk: {}", replicasOnFailedDisks);
+
+      long startTime = System.currentTimeMillis();
+      boolean success = false;
+      DistributedLock lock = primaryClusterParticipant.getDistributedLock("DISK_FAILURE", "Lock for disk failure");
+      while (true) {
+        if (!lock.tryLock()) {
+          // sleep for a while and try to acquire lock again
+          logger.info("Fail to acquire lock when handling disk failure, backoff some time and retry");
+          storeMainMetrics.handleDiskFailureRetryLockCount.inc();
+          backoff();
+          continue;
+        }
+        boolean inMaintenanceMode = false;
+        try {
+          // 1. enter maintenance mode
+          inMaintenanceMode = enterMaintenance();
+          // 2. remove all the replicasOnFailedDisks from the property store
+          removeReplicasFromCluster(replicasOnFailedDisks);
+          // 3: update disk availability
+          setDiskUnavailable(newFailedDisks);
+          // 4. reset partitions, we have to update disk availability before reset the partitions. This way, we know
+          // The partitions won't be re-assigned back to the same disks.
+          resetPartitions(replicasOnFailedDisks);
+          // 5. update disk capacity
+          updateDiskCapacity(healthyDiskCapacity);
+          // 6. Remove disks from the maps.
+          cleanupDisksAndReplicas(newFailedDisks, replicasOnFailedDisks);
+          logger.info("Successfully remove failed disks and replicas from memory when handling disk failure");
+          success = true;
+        } catch (Exception e) {
+          success = false;
+          storeMainMetrics.handleDiskFailureErrorCount.inc();
+        } finally {
+          // 7. exist maintenance mode
+          if (inMaintenanceMode) {
+            if (primaryClusterParticipant.exitMaintenanceMode()) {
+              logger.info("Successfully exit maintenance mode");
+            } else {
+              success = false;
```

Comment:
Can we add logger.error if "existMaintenanceMode" fails?

Reply:
HelixParticipant would log the error.